app/javascript/shared/components/useKeyboardShortcuts.jsx

```diff
@@ -0,0 +1,97 @@
+import { useEffect } from "preact/hooks";
+import PropTypes from 'prop-types';
+import { h } from 'preact';
+
+/**
+ * hook that can be added to a component to listen
+ * for keyboard presses
+ *
+ * @example
+ * const shortcuts = {
+ *   "ctrl+alt+KeyG": (e) => {
+ *     e.preventDefault();
+ *     alert("Control Alt G has been pressed")
+ *   }
+ * }
+ *
+ * useKeyboardShortcuts(shortcuts, someElementOrWindowObject);
+ *
+ * @param {object} shortcuts List of keyboard shortcuts/event
+ * @param {EventTarget} [eventTarget=window] An event target.
+ *
+ */
+export function useKeyboardShortcuts(shortcuts, eventTarget = window) {
+  useEffect(() => {
+    if (!shortcuts) return;
+
+    // Return true if element is a form element
+    const isFormField = (element) => {
+      if (!(element instanceof HTMLElement)) {
+        return false;
+      }
+
+      const name = element.nodeName.toLowerCase();
+      const type = (element.getAttribute("type") || "").toLowerCase();
+      return (
+        name === "select" ||
+        name === "textarea" ||
+        (name === "input" && ["submit", "reset", "checkbox", "radio"].indexOf(type) < 0) ||
+        element.isContentEditable
+      );
+    };
+
+    const keyEvent = (e) => {
```

Comment:
refactor (non-blocking): `keyEvent` depends on the argument `shortcuts`, but it does not need to be directly in the `useEffect`, it can be declared outside of the `useEffect` call but still reside in the `useKeyboardShortcuts` function body.

Reply:
Don't useEffects prefer a used function to be stated inside the useEffect where possible? If you have it outside, as a const for instance, doesn't that get reset on each render meaning the useEffect is recalled each render?

Reply:
Use effect will only be called when the dependency array is changed. The functions I'm suggesting to move out would be referenced only and created one time (when the ES module is first instantiated. If they remain in the `useEffect` callback, these functions will be recreated every time the `useEffect` callback is called.

Reply:
Maybe this is my understanding breaking down (or preact differs from React) but I thought.

```jsx
const FakeComponent = ({...props}) => {
  const someFunction = () => {
    // I'm created on every render
  }

  useEffect(()=>{
    someFunction();
    // therefore I run on every render
  },[props.cheese, someFunction])

  return null
}
```

compared to
```jsx
const FakeComponent = ({...props}) => {
  useEffect(()=>{
     const someFunction = () => {
       // I'm created every time the cheese prop changes
     }

    someFunction();
    // therefore I run on every cheese prop change.
  },[props.cheese])

  return null
}
```
and in React I'd do
```jsx
const FakeComponent = useCallback(({...props}) => {
  const someFunction = () => {
    // I'm created the first render only
  }, [])

  useEffect(()=>{
    someFunction();
    // therefore I run on every cheese prop change
  },[props.cheese, someFunction])

  return null
}
```

Reply:
You don't even need to use `useCallback`. What I meant was, you can write it like this.

```jsx
const someFunction = () => {
    // I'm declared only once
}

const FakeComponent = ({...props}) => {
  useEffect(()=>{
    someFunction(); // I'm called in use effect, but I'm never redeclared.
    // therefore I run on every render
  }, [props.cheese, someFunction])

  return null
}
```

`props.cheese` 😆

Reply:
The only time you need to have a function like `someFunction` declared within the function component is if it requires access to variables in the [closure](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures).

Reply:
In this bit of code the function `keyEvent` relies on `shortcuts`, so needs to stay somewhere where that is accessible. Unless you'd rather we pass shortcut from useEffect to the function outside?

```jsx
const someFunction = (cheese) => {
    // I'm declared only once
}

const FakeComponent = ({...props}) => {
  useEffect(()=>{
    someFunction(props.cheese); // I'm called in use effect, but I'm never redeclared.
    // therefore I run on cheese prop update
  }, [props.cheese])

  return null
}
```

Reply:
Ahh, I must have missed that. My bad. In that case I'd say pass it in, but I'll leave that up to you to decide.